### PR TITLE
[SROA] Use !tbaa instead of !tbaa.struct if op matches field.

### PR DESCRIPTION
**Pull request by fhahn**
If a split memory access introduced by SROA accesses precisely a single field of the original operation's !tbaa.struct, use the !tbaa tag for the accessed field directly instead of the full !tbaa.struct.

InstCombine already had a similar logic.

Motivation for this and follow-on patches is to improve codegen for libc++, where using memcpy limits optimizations, like vectorization for code iteration over std::vector<std::complex<float>>: https://godbolt.org/z/f3vqYos3c

Depends on https://github.com/llvm/llvm-project/pull/81285.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/SROA.cpp [hunk 1]
```diff
@@ -4561,6 +4577,10 @@ bool SROA::presplitLoadsAndStores(AllocaInst &AI, AllocaSlices &AS) {
         PStore->copyMetadata(*SI, {LLVMContext::MD_mem_parallel_loop_access,
                                    LLVMContext::MD_access_group,
                                    LLVMContext::MD_DIAssignID});
+
+        if (AATags)
+          PStore->setAAMetadata(
```
Comment (dobbelaj-snps) [inline llvm/lib/Transforms/Scalar/SROA.cpp:4582]: Is this part covered in a testcase ?

File: llvm/test/Transforms/SROA/tbaa-struct3.ll [hunk 2]
```diff
@@ -7,9 +7,9 @@ define void @load_store_transfer_split_struct_tbaa_2_float(ptr dereferenceable(2
 ; CHECK-NEXT:  entry:
 ; CHECK-NEXT:    [[TMP0:%.*]] = bitcast float [[A]] to i32
 ; CHECK-NEXT:    [[TMP1:%.*]] = bitcast float [[B]] to i32
-; CHECK-NEXT:    store i32 [[TMP0]], ptr [[RES]], align 4
+; CHECK-NEXT:    store i32 [[TMP0]], ptr [[RES]], align 4, !tbaa.struct [[TBAA_STRUCT0:![0-9]+]]
```
Comment (dobbelaj-snps) [inline llvm/test/Transforms/SROA/tbaa-struct3.ll:10]: Shouldn't we aim for also getting a !tbaa [[tbaa1]] here ? This is maybe related to my comment in #81285

**Comment by dobbelaj-snps** [comment]
Hmm. 10 changes + 1 new usage of setAAMetaData, But only 4 relevant changes in tests.. That part of SROA might seems to lack some testing ?

File: llvm/test/Transforms/SROA/tbaa-struct3.ll (hunk 2 above)
Comment (fhahn) [reply llvm/test/Transforms/SROA/tbaa-struct3.ll:10]: Yes, this should be solved by a separate improvement: #81313

**Comment by fhahn** [comment]
> Hmm. 10 changes + 1 new usage of setAAMetaData, But only 4 relevant changes in tests.. That part of SROA seems to lack some testing ?

Yes, will add the missing coverage, just wanted to make sure this makes sense in general beforehand

**Comment by dobbelaj-snps** [comment]
> > Hmm. 10 changes + 1 new usage of setAAMetaData, But only 4 relevant changes in tests.. That part of SROA seems to lack some testing ?
> 
> Yes, will add the missing coverage, just wanted to make sure this makes sense in general beforehand

I think it makes sense.

For this one to go through, I think we need at least a testcase that triggers the new 'PStore->setAAMetaData'... case. (line 4582)

File: llvm/lib/Transforms/Scalar/SROA.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Scalar/SROA.cpp:4582]: Should be now, I added a number of additional tests that should cover all cases here

**Review by fhahn — COMMENTED** [review]
@dobbelaj-snps Added a substantial number of tests that should cover all cases now in 2a9b86cc10c3883cca51a5166aad6e2b755fa958

File: llvm/lib/Transforms/Scalar/SROA.cpp (hunk 1 above)
Comment (dobbelaj-snps) [reply llvm/lib/Transforms/Scalar/SROA.cpp:4582]: thx !

**Review by dobbelaj-snps — APPROVED** [review]
lgtm

**Pull request merged**